File: 2024/day-01-historian-hysteria.go
package adventofcode

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"sort"
)

// Day 1: Historian Hysteria
// https://adventofcode.com/2024/day/1

func Day01Part1() error {
	filename := "day-01-historian-hysteria.data"
	file, err := os.OpenFile(filename, os.O_RDONLY, 0644)
	if err != nil {
		return errors.New("os.OpenFile ERROR: " + err.Error())
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)
	var first, second int
	var list1 = make([]int, 0)
	var list2 = make([]int, 0)
	for scanner.Scan() {
		_, err := fmt.Sscanf(scanner.Text(), "%d   %d", &first, &second)
		if err != nil {
			return errors.New("fmt.Sscanf ERROR: " + err.Error())
		}
		list1 = append(list1, first)
		list2 = append(list2, second)
	}
	sort.Ints(list1)
	sort.Ints(list2)
	var abs = func(n int) int {
		if n < 0 {
			return -n
		}
		return n
	}
	var total, distance = 0, 0
	for idx := range list1 {
		distance = abs(list1[idx] - list2[idx])
		total += distance
	}
	fmt.Println("total distance:", total)
	return nil
}

func Day01Part2() error {
	filename := "day-01-historian-hysteria.data"
	file, err := os.OpenFile(filename, os.O_RDONLY, 0644)
	if err != nil {
		return errors.New("os.OpenFile ERROR: " + err.Error())
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)
	var first, second int
	var list1 = make([]int, 0)
	var freq2 = make(map[int]int)
	for scanner.Scan() {
		_, err := fmt.Sscanf(scanner.Text(), "%d   %d", &first, &second)
		if err != nil {
			return errors.New("fmt.Sscanf ERROR: " + err.Error())
		}
		list1 = append(list1, first)
		freq2[second]++
	}
	var score, similarity = 0, 0
	for _, num := range list1 {
		similarity = num * freq2[num]
		score += similarity
	}
	fmt.Println("similarity score:", score)
	return nil
}

File: 2024/day-01-historian-hysteria.py
from collections import defaultdict
import unittest

# Day 1: Historian Hysteria
# https://adventofcode.com/2024/day/1

# python3 -m unittest 2024/day-01-historian-hysteria.py

class Solution(unittest.TestCase):
    filename = "2024/day-01-historian-hysteria.data"

    def part1(self) -> None:
        list1 = []
        list2 = []
        with open(self.filename, "r") as file:
            for line in file:
                nums = line.split()
                list1.append(int(nums[0]))
                list2.append(int(nums[1]))
        list1.sort()
        list2.sort()
        total = 0
        for idx in range(len(list1)):
            distance = abs(list1[idx] - list2[idx])
            total += distance
        print("total distance:", total)

    def part2(self) -> None:
        list1 = []
        count2 = defaultdict(int)
        with open(self.filename, "r") as file:
            for line in file:
                nums = line.split()
                list1.append(int(nums[0]))
                count2[int(nums[1])] += 1
        score = 0
        for num in list1:
            similarity = num * count2[num]
            score += similarity
        print("similarity score:", score)

    def test(self) -> None:
        self.part1()
        self.part2()
